src/sysvinit/inittab.5
.Dd May 15, 2015
.Dt INITTAB 5
.Os
.Sh NAME
.Nm inittab
.Nd describe the system runlevels
.Sh DESCRIPTION
The
.Pa /etc/inittab
file is parsed by
.Xr init 8
in order to figure out which processes to run for every runlevel.
An entry in an
.Nm
file consists of a line in the following format:
.Pp
.D1 Em id : Ns Em runlevels : Ns Em action : Ns Em command
.Pp
Empty lines, lines containing only whitespace, lines starting with
.Ic #
and invalid entries are ignored.
The meaning of each field is as follows:
.Pp
.Bl -tag -width runlevels
.It Em id
A unique identifier for this entry.
This field is not actually used, but required for legacy compatibility reasons.
.It Em runlevels
The runlevels this entry belongs to.
May contain multiple runlevels, such as
.Ic 016
for runlevels 0, 1 and 6.
Refer to
.Xr init 8
for details about runlevels.
.It Em action
The kind of action that should be taken for this entry.
Supported actions are:
.Bl -tag -width initdefault
.It Sy off
Do nothing for this entry.
All other fields are ignored.
.It Sy initdefault
The
.Em runlevels
field specifies the default runlevel the system starts as, if none is given otherwise.
The
.Em command
field is ignored.
.It Sy sysinit
The
.Em command
is ran before
.Sy boot
and
.Sy bootwait
entries;
.Cm init
waits for the command to exit before proceeding.
The
.Em runlevels
field is ignored.
.It Sy boot
The
.Em command
is ran before
.Cm init
switches to the initial runlevel.
The command is not waited for.
The
.Em runlevels
field is ignored.
.It Sy bootwait
Same as
.Sy boot ,
except that
.Cm init
waits for the command to exit before proceeding.
.It Sy once
The
.Em command
is ran when
.Cm init
switches to one of the runlevels in the
.Em runlevels
field.
The command is not waited for.
.It Sy wait
Same as
.Sy once ,
except
.Cm init
waits for the command to exit before proceeding.
.It Sy respawn
Same as
.Sy once ,
except
.Cm init
re-executes the command when it notices it has exited.
.El
.It Em command
The command to run for this entry, in the form of
.Cm /bin/command ARG1 ... ARGN .
This command will be passed to
.Cm /bin/sh Fl c
to execute.
.El
.Sh EXAMPLES
A minimal inittab using a separate service manager located at
.Pa /sbin/rc :
.Bd -literal -offset Ds
si::sysinit:/sbin/rc sysinit
bo::boot:/sbin/rc boot
bw::bootwait:/sbin/rc bootwait

0:0:wait:/sbin/rc 0
1:1:wait:/sbin/rc 1
3:3:wait:/sbin/rc 3
6:6:wait:/sbin/rc 6
.Ed
.Pp
A more elaborate example with TTY spawning:
.Bd -literal -offset Ds
# Run in single-user mode by default.
id:1:initdefault:

# Initialization.
si::sysinit:/sbin/rc sysinit
bo::boot:/sbin/rc boot
bw::bootwait:/sbin/rc bootwait

# Runlevels.
1e:1:once:echo Switching to runlevel 1...
1:1:wait:/sbin/rc 1
3e:3:once:echo Switching to runlevel 3...
3:3:wait:/sbin/rc 3

# Shutdown.
0:0:wait:/sbin/rc 0
6:6:wait:/sbin/rc 6

# Our 6 TTYs.
tty1:13:respawn:/sbin/agetty 38400 tty1
tty2:13:respawn:/sbin/agetty 38400 tty2
tty3:13:respawn:/sbin/agetty 38400 tty3
tty4:13:respawn:/sbin/agetty 38400 tty4
tty5:13:respawn:/sbin/agetty 38400 tty5
tty6:13:respawn:/sbin/agetty 38400 tty6
.Ed
.Sh AUTHORS
.An Shiz Aq Mt hi@shiz.me
.Sh SEE ALSO
.Xr init 8 ,
.Xr sh 1


src/sysvinit/inittab.c
#define  _BSD_SOURCE
#define  _DEFAULT_SOURCE
#include <unistd.h>
#include <string.h>
#include <fcntl.h>
#include "init.h"
#include "inittab.h"


/* parse inittab file */
int parse_tab(const char *filename, struct init_task *tasks, unsigned ntasks)
{
    int fd = open(filename, O_RDONLY);
    if (fd < 0)
        goto err;

    char buf[8192];
    int ret, off = 0;
    unsigned task = 0;
    while ((ret = read(fd, buf + off, sizeof(buf) - off - 1)) > 0) {
        buf[off + ret] = 0;
        off = 0;

        /* convert lines to separate strings */
        for (char *p = buf; *p; p++)
            if (*p == '\n')
                *p = 0;

        /* parse line by line */
        int parsed = 0;
        char *p = buf;
        while (parsed < ret) {
            char *line = p;
            int len = strlen(line);

            /* mark line as parsed */
            parsed += len + 1;
            p += len + 1;

            /* skip empty and comment lines */
            for (; *line; line++)
                if (*line == '#') {
                    while (*line)
                        line++;
                    break;
                } else if (*line != '\t' && *line != ' ')
                    break;

            if (!*line)
                continue;

            /* get fields */
            char *eid      = strsep(&line, ":");
            char *erunlvls = strsep(&line, ":");
            char *eaction  = strsep(&line, ":");
            char *ecmd     = line;

            if (!eid || !erunlvls || !eaction || !ecmd) {
                /* required fields not found, try to restart parsing if at end of buffer */
                char *linestart = p - len - 1;
                if (parsed >= ret && buf != linestart) {
                    memmove(buf, linestart, len + 1);
                    off = len;
                    continue;
                }
                goto err;
            }

            /* semantic checks */
            if (!*eid || !*eaction)
                goto err;

            int action = parse_action(eaction);
            if (!action)
                continue;

            if (!*erunlvls && action != ACTION_BOOT && action != ACTION_BOOTWAIT && action != ACTION_SYSINIT)
                goto err;

            int runlevels = parse_runlevels(erunlvls);
            if (!runlevels && action != ACTION_BOOT && action != ACTION_BOOTWAIT && action != ACTION_SYSINIT)
                continue;

            if (!*ecmd && action != ACTION_INITDEFAULT)
                goto err;

            /* add task if we have room for it */
            if (task < ntasks) {
                strncpy(tasks[task].id,  eid,  sizeof(tasks[task].id) - 1);
                strncpy(tasks[task].cmd, ecmd, sizeof(tasks[task].cmd) - 1);
                tasks[task].id[sizeof(tasks[task].id) - 1] = 0;
                tasks[task].runlevels = runlevels;
                tasks[task].action = action;
                tasks[task].cmd[sizeof(tasks[task].cmd) - 1] = 0;
                tasks[task].pid = -1;
                tasks[task].flags = 0;
            }
            task++;
        }
    }
    if (off)
        goto err;
    if (ret < 0)
        goto err;

    close(fd);
    return task;

err:
    if (fd >= 0)
        close(fd);
    return -1;
}

int parse_runlevels(const char *runlevels)
{
    int res = 0;

    for (; *runlevels; runlevels++)
        if (*runlevels == '0')
            res |= RUNLEVEL_SHUTDOWN;
        else if (*runlevels == '1')
            res |= RUNLEVEL_SINGLE;
        else if (*runlevels == '3')
            res |= RUNLEVEL_MULTI;
        else if (*runlevels == '6')
            res |= RUNLEVEL_REBOOT;

    return res;
}

int parse_action(const char *action)
{
    int res = 0;

    if (!strcmp(action, "off"))
        res = ACTION_OFF;
    else if (!strcmp(action, "initdefault"))
        res = ACTION_INITDEFAULT;
    else if (!strcmp(action, "sysinit"))
        res = ACTION_SYSINIT;
    else if (!strcmp(action, "boot"))
        res = ACTION_BOOT;
    else if (!strcmp(action, "bootwait"))
        res = ACTION_BOOTWAIT;
    else if (!strcmp(action, "once"))
        res = ACTION_ONCE;
    else if (!strcmp(action, "wait"))
        res = ACTION_WAIT;
    else if (!strcmp(action, "respawn"))
        res = ACTION_RESPAWN;

    return res;
}
